src/flint/types/fmpz_poly.pyx

```diff
@@ -344,18 +344,62 @@ cdef class fmpz_poly(flint_poly):
         return c, res
 
     def roots(self, bint verbose=False):
+        """
+        Computes all the integer roots of this polynomial.
+        Returns a list of all pairs (*v*, *m*) where *v* is the 
+        integer root and *m* is the multiplicity of the root
+
+            >>> fmpz_poly([]).roots()
+            []
+            >>> fmpz_poly([1]).roots()
+            []
+            >>> fmpz_poly([2, 1]).roots()
+            [(-2, 1)]
+            >>> fmpz_poly([1, 2]).roots()
+            []
+            >>> fmpz_poly([12, 7, 1]).roots()
+            [(-3, 1), (-4, 1)]
+            >>> (fmpz_poly([1, 2]) *  fmpz_poly([-3,1]) *  fmpz_poly([1, 2, 3]) *  fmpz_poly([12, 7, 1])).roots()
+            [(-3, 1), (-4, 1), (3, 1)]
+
+        """
+        cdef fmpz_poly_factor_t fac
+        cdef fmpz constant_coeff, linear_coeff
+        cdef long deg, i
+        cdef int exp
+
+        if not self:
+            return []
+
+        roots = []
+        constant_coeff = fmpz()        
+        linear_coeff = fmpz()
+
+        fmpz_poly_factor_init(fac)
+        fmpz_poly_factor(fac, self.val)
+        for 0 <= i < fac.num:
+            deg = fmpz_poly_degree(&fac.p[i])
+            fmpz_poly_get_coeff_fmpz(linear_coeff.val, &fac.p[i], 1)
```

Comment:
It would be much easier to implement this as just Python code:
```
In [26]: x = flint.fmpz_poly([0, 1])

In [27]: p = x*(x - 1)*(x**2 + 2)

In [28]: [(-fac[0], mul) for fac, mul in p.factor()[1] if fac.degree() == 1]
Out[28]: [(0, 1), (1, 1)]
```
Then it could be a generic method defined in the superclass.

Reply:
I kept this as cython in case it helped with performance but agree that we could instead write it in python this way. 

If the code was in the base superclass is the best thing then to have import statements in the doctest for each flint type?

For the arb and acb we can simply have roots write over the superclass for the specific impl which currently exist?

Reply:
Performance of anything else is unlikely to be a significant concern if the base routine is factor. Other algorithms could be much more efficient in common cases e.g. using the rational root theorem (and derivative for multiplicity):
https://en.wikipedia.org/wiki/Rational_root_theorem

Note also that Python code in a .pyx file will run faster than Python code in a .py file: Cython already translates the code to C to some extent even if it just looks like Python code.

If Flint had dedicated algorithms for this then it would be worth calling those from cython. Otherwise I don't think it is worth implementing anything nontrivial here and we should just try to keep the code here as simple as possible: it is better to spend time expanding the coverage of Flint's existing functionality than trying to optimise things that are not already optimised in Flint.

> For the arb and acb we can simply have roots write over the superclass for the specific impl which currently exist?

Yes, and if any specialised algorithms are implemented in future for other cases like `fmpz_poly` they can override the generic method as well.